### PR TITLE
Introduce configuration option to configure image pushing per kube-context

### PR DESCRIPTION
**Pull request by corneliusweig**
This PR adds a new configuration option in the skaffold configuration file called `local-cluster` with values true/false. It will be useful for users working with a local setup but with non-standard kube-context names (i.e. other than `minikube` or `docker-for-desktop`). The effect of this option is to skip pushing images when the cluster is considered local. It therefore addresses https://github.com/GoogleContainerTools/skaffold/issues/1158, https://github.com/GoogleContainerTools/skaffold/issues/1205, and https://github.com/GoogleContainerTools/skaffold/issues/1172.

I tried to make this new option fully backwards compatible. This required to make it optional, therefore it is implemented as a `*bool`. Unfortunately this breaks the config-set command, e.g. `skaffold config set --global local-cluster true`, because it cannot handle pointer types. Before this PR can be merged, this should be addressed (as well as adding some documentation). However, before moving on, I would like to get some feedback if you think this makes sense.

### CODE REVIEW

**Comment by corneliusweig** [comment]
I signed it!

**Comment by corneliusweig** [comment]
I signed it!

File: cmd/skaffold/app/cmd/config/set.go [hunk 1]
```diff
@@ -62,20 +77,16 @@ func setConfigValue(name string, value interface{}) error {
...
-	fieldType := fieldValue.Type()
-	val := reflect.ValueOf(value)
+	return fieldName
+}
 
-	if fieldType != val.Type() {
-		return fmt.Errorf("%s is not a valid value for field %s", value, fieldName)
+func getNewFieldValue(fieldName string, name string, value string) (reflect.Value, error) {
+	config := ContextConfig{}
+	auxiliaryContents := name + ": " + value
```
Comment (nkubala) [inline cmd/skaffold/app/cmd/config/set.go]: I'm not sure I understand why you need to do this. What was wrong with just doing a `reflect.ValueOf(newValue)`? Creating a new config and then doing `reflect.ValueOf(config).FieldByName(fieldName)` seems redundant to me, but I might be missing something.

I'm also not crazy about using anything yaml-related in this code: the entire reason I rewrote it to use reflection was to move away from using yaml marshaling/unmarshaling magic to set values in a struct.

File: cmd/skaffold/app/cmd/config/set.go [hunk 2]
```diff
@@ -46,12 +46,27 @@ func NewCmdSet(out io.Writer) *cobra.Command {
 	return cmd
 }
 
-func setConfigValue(name string, value interface{}) error {
+func setConfigValue(name string, value string) error {
```
Comment (nkubala) [inline cmd/skaffold/app/cmd/config/set.go]: The reason I originally had it as an `interface{}` was mostly for clarity that even if the provided value can be represented as a string, it might semantically be something different, i.e. a number, or even a nested struct. Is there any reason we need to make it a string here?

Comment (corneliusweig) [reply cmd/skaffold/app/cmd/config/set.go]: Right, that's a good reason.

In my new implementation, the type actually needs to be a string, because the `getNewFieldValue` function builds an auxiliary yaml from the input value.

Clearly, your previous version for `setConfigValue` is more generic. On the other hand, it doesn't handle type conversion of the given value into the desired target type. To me, adapting `setConfigValue` seemed to be the right place for the type conversion. But maybe the type conversion should be done in the caller of `setConfigValue`, then the original code would work just fine. What do you think?

File: cmd/skaffold/app/cmd/config/set.go (hunk 1 above)
Comment (corneliusweig) [reply cmd/skaffold/app/cmd/config/set.go]: So the reflection code was working very nicely with just `string` typed values. However, the new option is a `*bool`, so that the old code was trying to set it with a string value. This resulted in a type mismatch in line 73 old version, because `setConfigValue` is always called with actual type `string` for the value. But maybe I'm missing a simpler solution here.

I was very back and forth with another approach where the `LocalCluster` config is also made a `string`, so that the old code would have worked fine here. However, in the end I thought that there might be further settings in the future where such a workaround is even uglier than for `LocalCluster`.

**Comment by nkubala** [comment]
@corneliusweig sorry for the super late reply here. Thanks for the explanations, I understand your rationale now; sorry for being a little slow originally :) this isn't as simple as I thought it would be originally.

I'd really like to avoid using yaml parsing if we can here though; it has limitations that using reflection can help us get around (see https://github.com/GoogleContainerTools/skaffold/pull/872 for an example), especially when dealing with nested objects in the config.

I hacked together an alternative solution that involves switching on the fieldType of the config, and creating a new value object based on the string arg we pass in through the CLI args:

```golang
func setConfigValue(name string, value interface{}) error {
	cfg, err := getOrCreateConfigForKubectx()
	if err != nil {
		return err
	}

	fieldName := getFieldName(cfg, name)
	if fieldName == "" {
		return fmt.Errorf("%s is not a valid config field", name)
	}

	fieldValue := reflect.Indirect(reflect.ValueOf(cfg)).FieldByName(fieldName)
	fieldType := fieldValue.Type()
	val, err := getNewFieldValue(fieldType, value)
	if err != nil {
		return fmt.Errorf("%s is not a valid value for field %s", value, name)
	}

	reflect.ValueOf(cfg).Elem().FieldByName(fieldName).Set(val)

	return writeConfig(cfg)
}

func getNewFieldValue(fieldType reflect.Type, value interface{}) (reflect.Value, error) {
	valStr := value.(string)

	switch fieldType.String() {
	case "string":
		return reflect.ValueOf(valStr), nil
	case "*bool":
		if valStr == "" {
			return reflect.Zero(fieldType), nil
		}
		valBase, err := strconv.ParseBool(valStr)
		if err != nil {
			return reflect.Value{}, err
		}
		return reflect.ValueOf(&valBase), nil
	default:
		return reflect.Value{}, fmt.Errorf("unsupported type: %s", fieldType)
	}
}
```

It's a little less clean because it involves adding a new switch clause every time we add a field with a new type in the global config, but it gets around having to use yaml parsing to address the concept of a "nil" field value (e.g. `true`, `false`, and `nil`). I really feel like there's another solution for implementing "optional" values here using reflection, I'm just not seeing it.

Take a look at this and let me know what you think, in the meantime can you rebase your PR so CI passes? If we can't come up with anything better than we can decide between the two implementations and get this merged soon.

**Comment by corneliusweig** [comment]
Hi @nkubala,
no worries, better late than never ;). I better understand now, why you want to avoid the yaml unmarshalling. And I quite like your modifications. Apart from minor renamings, I used your patch as basis for a new commit.

My only worry is the cast `value.(string)` in set.go. Since `value` only has type constraint `interface{}`, this could blow up. However, it's in a private function and I have too little experience with this project to judge if this is acceptable. After all, you also explained that this is a deliberate choice to convey that the method is meant to not only work for strings.

Last, I would also like to add some documentation about the whole global config. What would be the best place for that?

**Comment by nkubala** [comment]
@corneliusweig yeah, after thinking about it some more I think it might actually make sense to just change the signature to `setConfigValue(name string, value string)`. my reasoning here is that the value we're passing in is coming from cobra args, which are **always** strings, so there's no real reason to call them `interface{}` if we're just gonna cast them back. the change I recommended still holds though, because even though we're *passing* the value as a string, semantically it really has an embedded type (e.g. `*bool`) which we can infer using reflection.

so basically what I'm saying is, you can go ahead and change all these functions to accept the value parameter as a string and then get rid of the typecast :) then this LGTM

**Comment by nkubala** [comment]
oh, and in response to the global config: I realized how utterly terrible the documentation for that feature is a few days ago, sorry about that 😨 if you'd like to add some, it probably fits best in the [`Concepts`](https://github.com/GoogleContainerTools/skaffold/blob/master/docs/content/en/docs/concepts/_index.md) section of the docs 👍

**Comment by corneliusweig** [comment]
Hi @nkubala , all done. I hope the documentation is somewhat helpful.
In addition, I also found what I think was an error in the documentation of the the `gcr.io` default-repo logic. Could you take a look?

File: docs/content/en/docs/concepts/_index.md [hunk 3]
```diff
@@ -23,13 +23,28 @@ read the configuration file from the current directory.
...
+| `test` |  Specifies how Skaffold should test artifacts. Skaffold supports [container-structure-tests](https://github.com/GoogleContainerTools/container-structure-test) to test built artifacts. See [Using testers](/docs/how-tos/testers) for more information. |
+| `deploy` |  Specifies how Skaffold should deploy artifacts. Skaffold supports using `kubectl`, Helm, or kustomize to deploy artifacts. See [Using Deployers](/docs/how-tos/deployers) for more information. |
 | `profiles`|  Profile is a set of settings that, when activated, overrides the current configuration. You can use Profile to override the `build` and the`deploy`> section. |
 
 You can learn more about the syntax of `skaffold.yaml` at
 [`skaffold.yaml References`](https://github.com/GoogleContainerTools/skaffold/blob/master/examples/annotated-skaffold.yaml).
 
+## Global configuration (~/.skaffold/config)
+
+Some context specific settings can be configured in the global configuration file `~/.skaffold/config`. You can configure the options for specific or any context.
```
Comment (nkubala) [inline docs/content/en/docs/concepts/_index.md]: This file location is just the default, but it can theoretically be anywhere. Let's change this to something like

```Some context specific settings can be configured in a global configuration file, defaulting to `~/.skaffold/config`.```

Comment (nkubala) [reply docs/content/en/docs/concepts/_index.md]: Also, last line:

```Options can be configured globally or for specific contexts.```

File: docs/content/en/docs/concepts/_index.md [hunk 4]
```diff
@@ -23,13 +23,28 @@ read the configuration file from the current directory.
...
+The options are
+
+| Option | Type | Description |
+| ------ | ---- | ----------- |
+| `default-repo` | string | The image registry where images are published (See below). |
+| `local-cluster` | boolean | If true, do not try to push images after building. By default, contexts with names `docker-for-desktop`, `docker-desktop`, or `minikube` are treated as local. |
+
+For example, to treat any context as local by default:
+```
+skaffold config set --global local-cluster true
```
Comment (nkubala) [inline docs/content/en/docs/concepts/_index.md]: :+1:

File: docs/content/en/docs/concepts/_index.md (hunk 3 above)
Comment (corneliusweig) [reply docs/content/en/docs/concepts/_index.md]: Sounds much better :+1:

**Comment by haf** [comment]
Also, do you have any idea as to why when running minikube with kvm or virtualbox, the docker daemon isn't properly configured to have access? And even if adding a service-account _json_key to the docker daemon in the virtual machine / guest, skaffold is not allowed to do the `pull`, but will reuse images from the guest docker daemon when those have been pulled manually.

I'm commenting on this PR since it would seem that you're working on unifying how the context is managed in Skaffold.

**Comment by corneliusweig** [comment]
@haf It's impossible to really help you, because there is not enough context. For example, have you done the `docker login` _after_ `source <(minikube docker-env)` and so on. Please take the time to collect all information and open a separate issue.

**Comment by haf** [comment]
@corneliusweig I've set up ALL logins possible before starting skaffold. I don't run `source <(minikube docker-env)` at all.

**Comment by corneliusweig** [comment]
@haf So I guess you want to use a registry with access restrictions during your docker builds. That will require the following steps:
```
minikube start
source <(minikube docker-env) # this will set up the minikube docker daemon in your current shell
docker login ... # this will login in the minikube docker daemon
kubectl config use-context minikube
skaffold dev
```
Does that work for you?

**Comment by haf** [comment]
@corneliusweig No that doesn't work, because despite being able to PULL from within minikube, when skaffold runs docker it doesn't authenticate successfully.
